fms/utils/gptq.py

```diff
@@ -174,3 +175,175 @@ def shard_gptq_linear(
+                ),
+                dtype=torch.int32,
+            ),
+        )
+        self.register_buffer(
+            "scales",
+            torch.zeros(
+                (math.ceil(in_features / self.group_size), out_features),
+                dtype=torch.float16,
+            ),
+        )
+        self.register_buffer("g_idx", torch.zeros(in_features, dtype=torch.int32))
+        if bias:
+            self.register_buffer(
+                "bias", torch.zeros((out_features), dtype=torch.float16)
+            )
+        else:
+            self.bias = None
+
+    def unpack_qparam(
+        self,
+        qparam: torch.Tensor,
+        bits: int = 4,
+        transpose: bool = False,
+    ) -> torch.Tensor:
+        """Unpack GPTQ tensor of quantized qweight or qzeros, expanding
+        input matrix "32 // bits" times along originally packed dimension.
+
+        - qweights are packed by GPTQ along dim=0 and expanded row-wise
+        - qzeros are packed by GPTQ along dim=1 and expanded column-wise
+        (matrix is transposed before and after packing)
+
+        qweight packed size: [in_feat * bits // 32, out_feat]
+        qzeros packed size: [n_grp, out_feat * bits // 32]
+                            with n_grp = in_feat / grp_size
+
+        qweight unpacked size: [in_feat, out_feat]
+        qzeros unpacked size: [n_grp, out_feat]
+
+        Parameters
+        ----------
+        qparam : torch.Tensor
+            a packed quantization parameter (qweight or qzeros)
+        bits : int
+            number of quantization bits
+        transpose : bool
+            apply transpose before and after unpacking (needed by qzeros)
+
+        Return
+        ------
+        qparam_unpacked : torch.Tensor
+            the quantization parameter (qweight or qzeros) unpacked
+        """
+
+        if transpose:
+            qparam = qparam.t()
+
+        device = qparam.device
+        unpack_size = 32 // bits
+        qparam_unpacked = torch.zeros(
+            (qparam.size(0) * unpack_size, qparam.size(1)),
+            dtype=torch.int,
+            device=device,
+        )
+
+        for j in range(unpack_size):
+            qparam_unpacked[j::unpack_size, :] = (qparam >> (j * bits)) & 0xF
+        return qparam_unpacked.t() if transpose else qparam_unpacked
+
+    def dequantize(self, qweights_unpacked, scales):
+        """Dequantize unpacked weight tensor
+
+        qweights_unpacked: [in_feat, out_feat]
+        scales: [n_groups, out_feat]
+        qzeros: assume symmetric quantization => constant (8)
+        g_idx: activation reordering not supported
+        """
+
+        zp = 8
+        in_feat, _ = qweights_unpacked.size()
+        n_grp, _ = scales.size()
+        return torch.mul(
+            (qweights_unpacked - zp).to(torch.float16),
+            scales.repeat_interleave(in_feat // n_grp, dim=0),
+        )
+
+    def forward(self, x):
+        qweights_unpacked = self.unpack_qparam(self.qweight, self.bits)
```

Comment:
do we want to dequantize at every forward step? I assume ram is not as much of an issue on CPU v GPU, but I guess it matches the behavior of gpus

Reply:
yes, there may be a bit of a speedup to be had by dequantizing the weights once to FP16 and storing them, if decompressing the model to 4x its GPTQ size is not an issue on CPU. However, from what I observed (across non-extensive experiments), the runtime bottleneck is by far performing matmul on CPU, not weight dequantization. I would keep the current implementation, at least for now